Save and restore Categories cache when Activity is recreated after sy…

lemonboston opened the pull request:
…stem killed it. #339 

---

I've added saving and restoring `CategoriesCache` from `savedInstanceState` in `MicroFragmentHostActivity`. This is to handle the case when Activity would be restored after it had been killed, it wouldn't be needed for just configuration changes. I've just used the `savedInstanceState != null` check nevertheless, to not complicate the logic with extra flags. Since serialization doesn't really happen on config change, I think it's fine.

I haven't been able to test this real scenario though, https://github.com/schedjoules/android-event-discovery-sdk/issues/330 will be kept open for that. So this PR is for applying a 'blind' fix for that possible crash, so that we can move forward.

I did test the parcelling added in this PR by forcing it.

Commit:
Save and restore Categories cache when Activity is recreated after system killed it. #339

lemonboston commented:
@dmfs One more note to this: I know that the fact that we still need to save the cache to savedState even though we use `ViewModeel` kind of takes away the supposed simplicity of it. So I don't consider this an ideal solution, I mean even using `ViewModel`. I like you're ideas to handle states uniformly and go towards immutable views much better, too, so in the long run we can figure out architectures for that (I have some ideas already..).

But at this point this `ViewModel` for the categories cache looks good enough for me, to have a first version of the feature.

dmfs commented on eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/activities/MicroFragmentHostActivity.java:
```diff
@@ -87,9 +97,9 @@ protected void onSaveInstanceState(Bundle outState)
-        // TODO Possibly create mutable State abstraction that holds all state for the Activity.
-        // TODO Key should not be used like this:
-        outState.putParcelable(Keys.MICRO_FRAGMENT_HOST.name(), new ParcelableBox<>(mMicroFragmentHost));
+        new MutableBundle(outState)
```
Not a fan of this. IMO it's better/cleaner/simpler to use a `BundleBuilder` and add the result "manually" (as a nested bundle) to `outState` using `putParcelable`.
This `MutableBundle` thing doesn't really add anything and the name "MutableBundle" already indicates its pointlessness, because a `Bundle` is already mutable.

lemonboston replied:
So that could be something like this:
```java
        outState.putParcelable("schedjoules.state",
                new BundleBuilder()
                        .with(Keys.MICRO_FRAGMENT_HOST, new ParcelableBox<>(mMicroFragmentHost))
                        .with(Keys.CATEGORIES, new CategoriesBox(new CategoriesCache(this)))
                        .build());
```

My problem with this is that we start to expose the details of our serialization mechanism, I would like to avoid calling `putParcelable` directly or probably calling anything on `Bundle`.
So there is that `"schedjoules.state"` string key here now, which is not a `Key`. We can make it one, but then we would call `.name()` on it, which is not how it is supposed to work.

This is how we read the value in `onCreate()` now:
```java
new Argument<>(Keys.CATEGORIES, savedInstanceState).get()
```
This is straightforward, in-line with how we use it for Activity Intent or Fragment args. No need to think. With the version above, I wouldn't be sure how to access the values.

So I think there is great value in not exposing how we put and retrieve values to and from Bundles, especially nesting, and just using `Key<T>` accordingly.

`MutableBundle` is not the best name, I agree. The other option I thought is putting back that `Parameter.into(Bundle)` interface..  not sure if that's better. That couldn't be chained.

How about `FluentBundle` for the class? We are basically wrapping a Bundle here for adapting to our serialization and also making it fluent, so it seems fitting.

lemonboston commented:
I needed to use this in the filter ui branch as well, so I've renamed `MutableBundle` to `FluentBundle` and the method to `put` instead of `add` and rebased.

I think that how to do this may be part of a bigger topic of state saving, I have some ideas about that to discuss but I'd postpone that for later.

lemonboston commented:
I've rebased this onto `master`, we need this one in the next release. (Just found a way to test this accidentally during the regression test... it causes a crash)
